src/openms/source/FORMAT/MzTabM.cpp

```diff
@@ -0,0 +1,738 @@
+      {
+        ControlledVocabulary::CVTerm cvterm;
+        cvterm = cv.getTermByName("de novo search");
+        identification_method.fromCellString("[MS, " + cvterm.id + ", " + cvterm.name + ", ]");
+        id_mslevel = 2;
+      }
+      if (identification_software == "MetaboliteSpectralMatcher")
+      {
+        ControlledVocabulary::CVTerm cvterm;
+        cvterm = cv.getTermByName("TOPP SpecLibSearcher");
+        identification_method.fromCellString("[MS, " + cvterm.id + ", " + cvterm.name + ", ]");
+        id_mslevel = 2;
+      }
+      ControlledVocabulary::CVTerm cvterm_level = cv.getTermByName("ms level");
+      ms_level.fromCellString("[MS, " + cvterm_level.id + ", " + cvterm_level.name + ", " + String(id_mslevel) + "]");
+    }
+    if (identification_method.isNull())
+    {
+      OPENMS_LOG_WARN << "The identification method of your computational analysis can not be assessed'.\n"
+                      << "Please check if the ProcessingActions are set correctly!" << std::endl;
+    }
+
+    // ####################################################
+    //
+    // ####################################################
+
+    int feature_section_entry_counter = 1;
+    int evidence_section_entry_counter = 1;
+    for (auto& f : feature_map) // iterate over features and fill all sections
+    {
+      auto match_refs = f.getIDMatches();
+      if (match_refs.empty()) // features without identification
+      {
+        MzTabMSmallMoleculeFeatureSectionRow smf;
+        smf.smf_identifier = MzTabString(feature_section_entry_counter);
+        std::vector<MzTabString> corresponding_evidences;
+        smf.sme_id_refs.setNull(true);
+        if (f.metaValueExists("adducts"))
+        {
+          StringList adducts = f.getMetaValue("adducts");
+          smf.adduct = MzTabString(ListUtils::concatenate(adducts,'|'));
+        }
+        else
+        {
+          smf.adduct.setNull(true);
+        }
+        smf.sme_id_ref_ambiguity_code.setNull(true);
+        smf.isotopomer.setNull(true);
+        smf.exp_mass_to_charge = MzTabDouble(f.getMZ());
+        smf.charge = MzTabInteger(f.getCharge());
+        smf.retention_time = MzTabDouble(f.getRT());
+        smf.rt_start.setNull(true); // TODO: how to get that information
+        smf.rt_end.setNull(true); // TODO: how to get that information
+        smf.small_molecule_feature_abundance_assay[1] = MzTabDouble(f.getIntensity()); // only one map in featureXML
+
+        addMetaInfoToOptionalColumns(feature_user_value_keys, smf.opt_, String("global"), f);
+
+        smfs.emplace_back(smf);
+        ++feature_section_entry_counter;
+      }
+      else
+      {
+        // feature row based on number of individual identifications and adducts!
+        std::map<String, std::vector<int>> evidence_id_ref_per_adduct;
+        for (const IdentificationDataInternal::ObservationMatchRef& match_ref : match_refs) // iterate over all identifications of a feature
```

Comment:
IMHO, the sorting of `IdentificationDataInternal::ObservationMatchRef&` should take place before the for loop. 

I tried:
`std::sort(match_refs.begin(), match_refs.end(), MzTabM::myCompare_);`

```
bool myCompare_(const IdentificationDataInternal::ObservationMatchRef& lhs, const IdentificationDataInternal::ObservationMatchRef& rhs) {
    return std::tie(lhs->identified_molecule_var.getIdentifiedCompoundRef()->identifier) < std::tie(rhs->identified_molecule_var.getIdentifiedCompoundRef()->identifier);
  }
```

But it does not seem to be compatible with the boost structure. Any ideas, suggestions here?

Reply:
The issue 
is that the operator< is defined as the Address - which is not stable: https://github.com/OpenMS/OpenMS/blob/7f5009cfbabe74c126eece86f86cb9b7a5e4af62/src/openms/include/OpenMS/METADATA/ID/MetaData.h#L51

typedef IteratorWrapper<ObservationMatches::iterator> ObservationMatchRef;

One way this could be done is by adding a template specialization on ObservationMatches::iterator to IteratorWrapper that defines the operator< as you did. But this will probably not work with python. So I would suggest to create a 
```
struct ObservationMatchRef : public ObservationMatches::iterator
{
  bool operator<  ...
}

this should be wrappable in pyOpenMS and then provide your comparison operator

Reply:
I think we need to define both operator< and the equality operator in a meaningful way. Otherwise we can't properly compare ID data structures.

Reply:
I mean for IdentificationData as a whole as well. So basically we can check if two ID are the same (everything except for the Addresses)

Reply:
Thanks for the help! 
In terms of general comparison operators - I think that goes a bit too far for this PR - probably should be added to an Issue?

Reply:
I have it as an unchecked box in hendriks PR so we kind of know that it (likely) needs to be done